### PR TITLE
Closes #1136: Do not persist transient errors in caches

### LINKED ISSUE #1136: Do not persist transient errors in caches
Currently all exceptions occurring during patent metadata retriever communication with remote EPO endpoint, which are being handled by `PatentMetadataRetrieverJob#getMetadataFromFacade()`, are persisted within the cache alongside an empty patent content. Similarly in `CachedWebCrawlerJob`, responsible for web pages retrieval, any error is stored within the cache.

This is an expected behaviour for persistent errors but not for transient ones. Those temporary errors may occur e.g. when remote EPO (or software repository) endpoint rate limits the client quite severely resulting in exceeding the retries limit (set to 20 by default with 10 secs wait interval between the retries).

We should distinguish one type of errors from the other and whenever an exception is a result of temporary problems we should return the appropriate fault and empty patent metadata contents, as we do it right now, but we **should not** persist this pair within the cache. It should be possible to retry the patent retrieval in the next IIS round.

### PR DESCRIPTION
**Pull request by przemyslawjacewicz**
This PR removes caching of transient errors for patent metadata retrieval job and cached webcrawler job.

Summary of changes
1. new service facade response model in `FacadeContentRetrieverResponse` - to implement the core part of the task, that is division of persistent and transient errors for content retrieval from service I've created a simple class hierarchy
   * `FacadeContentRetrieverResponse.Success` - class wrapping the successful response from the facade
   * `FacadeContentRetrieverResponse.PersistentFailure` - wrapping a persistent error response from the facade 
   * `FacadeContentRetrieverResponse.TransientFailure` - wrapping a transient error response from the facade 

I've changed that because I wanted to make a simple correspondence of service facade response to a successful response from the service or a failed response from the service. Previous response model did not allowed for this - the response had a content and null exception or an empty content and an exception. Having a hierarchy of response classes also makes filtering of facade responses easier. This new model forced a few changes in service facade implementations.  

1. I've unified service facades implementations in `FacadeContentRetriever` to avoid unnecessary code duplication. This also made service facades not relying on extensive exception throwing for control flow. 

Minor things:
1. missing/new tests implemented
   * I've added junit5's `@DisplayName` annotation to make tests more clearly tell what is tested and how it is tested
1. compactions of factory classes for testing 
1. `ServiceFacadeUtils` updated - adding new utility method for factory creation using class name and parameters (to remove code duplication) 
1. `HdfsUtils` updated - adding method for counting files by extension

### CODE REVIEW

File: iis-wf/iis-wf-referenceextraction/src/main/java/eu/dnetlib/iis/wf/referenceextraction/softwareurl/HttpServiceFacade.java
Comment (marekhorst) [inline iis-wf/iis-wf-referenceextraction/src/main/java/eu/dnetlib/iis/wf/referenceextraction/softwareurl/HttpServiceFacade.java:126]: The comment related to patent retriever and `default` section is applicable here as well. Let's discuss whether we should use `persistentFailure()` or `transientFailure()` here.

**Review by marekhorst — COMMENTED** [review]
I've got some doubts, let's discuss it before final approval.

File: iis-common/src/main/java/eu/dnetlib/iis/common/java/io/HdfsUtils.java
Comment (przemyslawjacewicz) [reply iis-common/src/main/java/eu/dnetlib/iis/common/java/io/HdfsUtils.java:79]: We've got two methods for counting files and both are used only in tests. So I'm moving both to test hierarchy.

Comment (przemyslawjacewicz) [reply iis-common/src/main/java/eu/dnetlib/iis/common/java/io/HdfsUtils.java:79]: BTW I've created a ticket #1231 for replacing file counting using a path filter with file counting using file extension.

**Comment by przemyslawjacewicz** [comment]
Pushed updates after code review. To verify errors from the facade falling into 'default' case I'm running cached webcrawler and patent metadata retriever jobs with empty caches. Both are still running. When they finish I'll add the error breakdown to the related issue.

**Review by marekhorst — APPROVED** [review]
LGTM.

**Pull request merged**
